nanopub/client.py

```diff
@@ -245,7 +245,15 @@ def _search(self, endpoint: str, params: dict):
             else:
                 r = self._query_grlc(params, endpoint, grlc_url)
                 r.raise_for_status()
-            results = r.json()
+
+            # Check if JSON was actually returned. HTML can be returned instead
+            # if e.g. virtuoso errors on the backend (due to spaces in the search
+            # string, for example).
+            try:
+                results = r.json()
+            except ValueError:
```

Comment:
Why ValueError and not JSONDecodeError?

Reply:
Mainly because the online examples used it, possibly to avoid needing to consider what json lib ```requests``` is using. But since it is obviously going to use ```json``` from now on then we can use JSONDecodeError if you prefer?

I think I still like the idea of catching similar exceptions created during the conversion to JSON since a ```nanopub``` user will not understand or be able to do anything about that. Maybe we should dump it all to an error log that users can access if they want insight into why server queries are failing?

Reply:
This API seems a bit funny to me. Is it designed for human interaction, or for machine interaction?

If the former, then it should communicate in HTML, if the latter, then in JSON. But now it seems that it communicates data in machine-readable format if things are well (which is good if there's a script on the other side, and useless if there's a human), and in human-readable format if things are not well (which is good if there's a human on the other side, but useless to a script).

So it seems that the idea is that a script queries it, and if there's an error, fires up a browser to show a user what's wrong? Assuming that a browser makes sense in whatever environment the script was started in? What Python library does that?

The standard way of signalling an error in Python is by raising an exception. So that's what the nanopub library should do. And then I guess it just needs to do its best to give as good an error message as it can, given the limitations imposed by the server. If the HTML is simple, maybe you can strip the tags and put whatever remains in the exception message? Or parse it and extract a specific element?

Reply:
@LourensVeen there is nothing we can do about the nanopub server returning us HTML. But we currently return `None` to the user in that case which is strange. We should give a more informative error message.

@raar1 cool, let's keep using `ValueError`!

Reply:
I agree that it is ugly, but as Sven says we don't really have much control over this behaviour. I would like to convert the exception to a warning, effectively. Something along the lines of ```Warning: Server did not return JSON as expected. This query caused an error.```

Perhaps what we should do is catch the 'bad' text string before we send it? Ideally though, the server would always return JSON if asked for it, so we could instead see if the people running the different nanopub servers could agree on a standard response? I think in principle they might not all be using virtuoso as the triple store, so this 'free form text' error might not always happen...

Reply:
I quite like the message in that warning, but I would make it throw an actual error... I don't think we should do more than that.

Reply:
Done.